adr/ADR-50.md

```diff
@@ -91,7 +90,82 @@ The `LastMsgId` header is currently not supported. A batch will be rejected if t
+
+Crucially these batches are not pre-staged and so will not be limited to 1000 messages like Atomic ones.
+
+### Control Channel
+
+The heart of this design is control channel that is open and kept open for the duration of the batch. In practise this the reply subject of the first message will be used for all signaling during the batch.
+
+The server will send acks over the channel on a frequency like once every 10 messages or once every 5MB. Crucially if at any stage an error is encountered errors can be sent back immediately and received by the client.
+
+Clients will not wait for each ack like they would in a standard JS Publish instead they will maintain a count of maximum outstanding acknowledgements from the server, this is part flow-control and part outage detection.
+
+The server on the other hand will be able to adjust the frequency of acks based on internal metrics of the stream such as the size of the in-flight RAFT proposals. In this way the server and clients will balance speed of fast ingest vs responsiveness of interactive commands requiring RAFT proposals.
+
+At all times the server will maintain its self-protection mechanisms like dropping messages when buffers are full etc.
+
+### Client Design
+
+The client will signal batch start, membership and flow control characteristics using headers on published messages.
+
+ * The client will set up a Inbox subscription that will be used for the duration of the batch
+ * A batch will be started by adding the `Nats-Fast-Batch-Id:uuid`, `Nats-Flow:10` and `Nats-Batch-Sequence:1` headers using a *request* with reply being the above Inbox, the server will reply with error or zero byte message. Maximum length of the ID is 64 characters.
+ * Following messages in the same batch will include the `Nats-Fast-Batch-Id:uuid` header and increment `Nats-Batch-Sequence:n` by one, and might optionally include a reply set to the same Inbox to force a immediate zero byte reply
+ * If the final message has the headers `Nats-Batch-Id:uuid`, `Nats-Batch-Sequence:n` and `Nats-Batch-Commit:1`, the server will store the message, commit the batch and reply with a pub ack.
+ * Otherwise, the final message will have headers `Nats-Batch-Id:uuid`, `Nats-Batch-Sequence:n` and `Nats-Batch-Commit:eob` and the server will commit the batch without storing the message and reply with a pub ack.
+ * Clients will monitor the acks and should an ack have a `Nats-Flow` header that differs from the active one they will adjust accordingly
+
+The `Nats-Flow` message has a few formats see the later section.
+
+The server will acknowledge in the following manner:
+
+ * The initial message will get an error - for example, feature not supported - or zero byte ack
+ * The server will then send zero byte acks back based on the values of `Nats-Flow`, the `Nats-Flow` header will always be included, such acks can optionally include a header `Nats-Batch-Sequence:n` which indicates which message triggered the ack
```

Comment:
At this point I think we need a clearly defined Ack - possible variants, content, headers etc.
It's getting pretty complex.